File: LWThemeManager/Classes/LWThemeExtension.m
//
// Created by luowei on 2019/12/11.
//

#import "LWThemeExtension.h"


@implementation UIColor (LWTheme)

+ (UIColor *)theme_colorWithRGBAString:(NSString *)RGBAString {
    UIColor *color = nil;

    NSArray *rgbaComponents = [RGBAString componentsSeparatedByString:@","];
    float RED = 0.0f;
    float GREEN = 0.0f;
    float BLUE = 0.0f;
    float ALPHA = 0.0f;

    //string like : 127,127,127
    if ([rgbaComponents count] == 3) {
        RED = [(NSString *) rgbaComponents[0] floatValue] / 255;
        GREEN = [(NSString *) rgbaComponents[1] floatValue] / 255;
        BLUE = [(NSString *) rgbaComponents[2] floatValue] / 255;

        color = [UIColor colorWithRed:RED green:GREEN blue:BLUE alpha:1.0f];

        //string like : 127,127,127,255
    } else if ([rgbaComponents count] == 4) {
        RED = [(NSString *) rgbaComponents[0] floatValue] / 255;
        GREEN = [(NSString *) rgbaComponents[1] floatValue] / 255;
        BLUE = [(NSString *) rgbaComponents[2] floatValue] / 255;
        ALPHA = [(NSString *) rgbaComponents[3] floatValue] / 255;

        color = [UIColor colorWithRed:RED green:GREEN blue:BLUE alpha:ALPHA];
    }

    return color;
}

//从UIColor得到RGBA字符串
+ (NSString *)theme_rgbaStringFromUIColor:(UIColor *)color {

    if (!color) {
        return nil;
    }

    if (color == [UIColor whiteColor]) {
        // Special case, as white doesn't fall into the RGB color space
        return @"255,255,255,255";
    }

    CGFloat red;
    CGFloat blue;
    CGFloat green;
    CGFloat alpha;

    [color getRed:&red green:&green blue:&blue alpha:&alpha];

    int redDec = (int) (red * 255);
    int greenDec = (int) (green * 255);
    int blueDec = (int) (blue * 255);
    int alphaDec = (int) (alpha * 255);

    NSString *returnString = [NSString stringWithFormat:@"%d,%d,%d,%d", (unsigned int) redDec, (unsigned int) greenDec, (unsigned int) blueDec, (unsigned int) alphaDec];

    return returnString;

}



@end

File: LWThemeManager_swift/Classes/LWThemeExtension.swift
//
//  LWThemeExtension.swift
//  LWThemeManager
//
//  Swift version of theme extensions
//  Copyright (c) 2015 luowei. All rights reserved.
//

import UIKit

extension UIColor {

    /// Creates a UIColor from an RGBA string
    /// - Parameter rgbaString: String in format "r,g,b" or "r,g,b,a" where values are 0-255
    /// - Returns: UIColor instance or nil if string is invalid
    static func theme_color(withRGBAString rgbaString: String) -> UIColor? {
        let components = rgbaString.components(separatedBy: ",")

        // Format: "r,g,b"
        if components.count == 3 {
            guard let red = Float(components[0]),
                  let green = Float(components[1]),
                  let blue = Float(components[2]) else {
                return nil
            }

            return UIColor(
                red: CGFloat(red / 255.0),
                green: CGFloat(green / 255.0),
                blue: CGFloat(blue / 255.0),
                alpha: 1.0
            )
        }

        // Format: "r,g,b,a"
        if components.count == 4 {
            guard let red = Float(components[0]),
                  let green = Float(components[1]),
                  let blue = Float(components[2]),
                  let alpha = Float(components[3]) else {
                return nil
            }

            return UIColor(
                red: CGFloat(red / 255.0),
                green: CGFloat(green / 255.0),
                blue: CGFloat(blue / 255.0),
                alpha: CGFloat(alpha / 255.0)
            )
        }

        return nil
    }

    /// Converts UIColor to RGBA string
    /// - Returns: String in format "r,g,b,a" where values are 0-255
    func theme_rgbaString() -> String? {
        // Special case for white color
        if self == UIColor.white {
            return "255,255,255,255"
        }

        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0

        guard getRed(&red, green: &green, blue: &blue, alpha: &alpha) else {
            return nil
        }

        let redDec = Int(red * 255)
        let greenDec = Int(green * 255)
        let blueDec = Int(blue * 255)
        let alphaDec = Int(alpha * 255)

        return "\(redDec),\(greenDec),\(blueDec),\(alphaDec)"
    }
}

extension CGColor {

    /// Creates a CGColor from an RGBA string
    /// - Parameter rgbaString: String in format "r,g,b" or "r,g,b,a" where values are 0-255
    /// - Returns: CGColor instance or nil if string is invalid
    static func theme_color(withRGBAString rgbaString: String) -> CGColor? {
        return UIColor.theme_color(withRGBAString: rgbaString)?.cgColor
    }

    /// Converts CGColor to RGBA string
    /// - Returns: String in format "r,g,b,a" where values are 0-255
    func theme_rgbaString() -> String? {
        return UIColor(cgColor: self).theme_rgbaString()
    }
}
